### PR TITLE
Edns extras

### PR DESCRIPTION
**Pull request by vlevigneron**
- First implementations of extra test cases for EDNS (DNS Flag Day). Nameserver10, Nameserver11, Nameserver12, Nameserver13 and Nameserver14
- See https://github.com/zonemaster/zonemaster/pull/650 forNameserver10
- See https://github.com/zonemaster/zonemaster/pull/661 for Nameserver11
- See https://github.com/zonemaster/zonemaster/pull/658 for Nameserver12
- See https://github.com/zonemaster/zonemaster/pull/653 for Nameserver13
- See https://github.com/zonemaster/zonemaster/pull/654 for Nameserver14

### CODE REVIEW

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 1]
```diff
@@ -134,6 +140,22 @@ sub metadata {
               CASE_QUERIES_RESULTS_DIFFER
               )
         ],
+        nameserver10 => [
+            qw(
+              NO_RESPONSE
+              NO_EDNS_SUPPORT
+              BAD_UNSUPPORTED_VER
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Nameserver.pm]: BAD_UNSUPPORTED_VER --> UNSUPPORTED_EDNS_VER

See merged spec https://github.com/zonemaster/zonemaster/blob/develop/docs/specifications/tests/Nameserver-TP/nameserver10.md

**Comment by matsduf** [comment]
@vlevigneron, can you update the description of the PR so it reflects the extended contents (more test cases)?

Please explain the missing messages commit in this PR.

**Comment by vlevigneron** [comment]
@matsduf 'missing messages commit' is about new entries in profile.json test_levels. It has been re-ordered so that's why there are more changes that just the addition of the few entries related to Nameserver*.

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 2]
```diff
@@ -760,6 +818,324 @@ sub nameserver09 {
...
+                    NO_EDNS_SUPPORT => {
+                        ns      => $ns->name,
+                        address => $ns->address->short,
+                    }
+                  );
+            }
+            elsif ( $p->rcode eq q{NOERROR} and $p->edns_version == 0 and $p->edns_z == 0 and $p->get_records( q{SOA}, q{answer} ) ) {
+                next;
+            }
+            else {
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Nameserver.pm]: Missing test for unsupported flags.

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 3]
```diff
@@ -760,6 +818,324 @@ sub nameserver09 {
...
+            if ( $p->rcode eq q{FORMERR} ) {
+                push @results,
+                  info(
+                    NO_EDNS_SUPPORT => {
+                        ns      => $ns->name,
+                        address => $ns->address->short,
+                    }
+                  );
+            }
+
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Nameserver.pm]: Missing test for truncated message without ENDS.

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 4]
```diff
@@ -760,6 +818,324 @@ sub nameserver09 {
...
+                push @results,
+                  info(
+                    NO_EDNS_SUPPORT => {
+                        ns      => $ns->name,
+                        address => $ns->address->short,
+                    }
+                  );
+            }
+            elsif ( $p->rcode eq q{BADVERS} and $p->edns_version == 0 and not defined $p->edns_data and $p->get_records( q{SOA}, q{answer} ) ) {
+                next;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Nameserver.pm]: Some action here. I am not sure that Nameserver14 makes sense.

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 5]
```diff
@@ -134,6 +149,45 @@ sub metadata {
...
+              NO_EDNS_SUPPORT
+              UNKNOWN_OPTION_CODE
+              NS_ERROR
+              )
+        ],
+        nameserver12 => [
+            qw(
+              NO_RESPONSE
+              NO_EDNS_SUPPORT
+              NS_ERROR
```
[the remark this answers is not in the document]
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Nameserver.pm]: @matsduf Just for clarification, in the specs, it is written :

> Else, if the pseudo-section has an OPT record with flags being set to some value, then output Z_FLAGS_NOTCLEAR.

RFC specifie 2 flags, DO  and Z. When we talk about "flagS", do we only talk about Z flag only or both ? Please re-formulate.

File: lib/Zonemaster/Engine/Test/Nameserver.pm (hunk 4 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Nameserver.pm]: I will remove for the moment Nameserver14 from this PR.

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 6]
```diff
@@ -134,6 +149,45 @@ sub metadata {
...
+              MISSING_OPT_IN_TRUNCATED
+              )
+        ],
+        nameserver14 => [
+            qw(
+              NO_RESPONSE
+              NO_EDNS_SUPPORT
+              NS_ERROR
```
[the remark this answers is not in the document]
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Nameserver.pm]: I guess this is no longuer relevant as we will remove nameserver14 from this release.

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 7]
```diff
@@ -167,14 +221,19 @@ sub translation {
         IPV4_DISABLED              => 'IPv4 is disabled, not sending "{rrtype}" query to {ns}/{address}.',
         IPV6_DISABLED              => 'IPv6 is disabled, not sending "{rrtype}" query to {ns}/{address}.',
         IS_A_RECURSOR              => 'Nameserver {ns}/{address} is a recursor.',
+        MISSING_OPT_IN_TRUNCATED   => 'Nameserver {ns}/{address}...',
```
[the remark this answers is not in the document]
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Nameserver.pm]: Fixed.

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 8]
```diff
@@ -167,14 +221,19 @@ sub translation {
...
+        NO_EDNS_SUPPORT            => 'Nameserver {ns}/{address}...',
```
[the remark this answers is not in the document]
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Nameserver.pm]: Fixed.

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 9]
```diff
@@ -167,14 +221,19 @@ sub translation {
...
         NO_RESOLUTION              => 'No nameservers succeeded to resolve to an IP address.',
         NO_RESPONSE                => 'No response from {ns}/{address} asking for {dname}.',
         NO_UPWARD_REFERRAL         => 'None of the following nameservers returns an upward referral : {names}.',
+        NS_ERROR                   => 'Nameserver {ns}/{address}...',
         QNAME_CASE_INSENSITIVE     => 'Nameserver {ns}/{address} does not preserve original case of queried names.',
         QNAME_CASE_SENSITIVE       => 'Nameserver {ns}/{address} preserves original case of queried names.',
         QUERY_DROPPED              => 'Nameserver {ns}/{address} dropped AAAA query.',
         SAME_SOURCE_IP             => 'All nameservers reply with same IP used to query them.',
+        UNKNOWN_OPTION_CODE        => 'Nameserver {ns}/{address}...',
+        UNSUPPORTED_EDNS_VER       => 'Nameserver {ns}/{address}...',
```
[the remark this answers is not in the document]
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Nameserver.pm]: Fixed.

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 10]
```diff
@@ -167,14 +221,19 @@ sub translation {
...
         UPWARD_REFERRAL            => 'Nameserver {ns}/{address} returns an upward referral.',
         UPWARD_REFERRAL_IRRELEVANT => 'Upward referral tests skipped for root zone.',
```
[the remark this answers is not in the document]
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Nameserver.pm]: Fixed.

File: lib/Zonemaster/Engine/Test/Nameserver.pm (hunk 2 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Nameserver.pm]: See remarks on  Z_FLAGS_NOTCLEAR.

File: lib/Zonemaster/Engine/Test/Nameserver.pm (hunk 5 above)
[the remark this answers is not in the document]
Comment (matsduf) [reply lib/Zonemaster/Engine/Test/Nameserver.pm]: Yes, we refer to the Z flag bits. In zonemaster/zonemaster#662 I propose a clarification.

There are 16 bits for flags, only one flag is defined ([RFC 6891:6.1.3-4]). I think it is reasonable to see those 15 bits as 15 undefined flags.
```
   The extended RCODE and flags, which OPT stores in the RR Time to Live
   (TTL) field, are structured as follows:

                  +0 (MSB)                            +1 (LSB)
       +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
    0: |         EXTENDED-RCODE        |            VERSION            |
       +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
    2: | DO|                           Z                               |
       +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
(...)
   DO
      DNSSEC OK bit as defined by [RFC3225].

   Z
      Set to zero by senders and ignored by receivers, unless modified
      in a subsequent specification.
```

Also see [EDNS Header Flags] in IANA repository.

[EDNS Header Flags]: https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-13
[RFC 6891:6.1.3-4]: https://tools.ietf.org/html/rfc6891#section-6.1.3

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 11]
```diff
@@ -167,16 +211,22 @@ sub translation {
...
+        NS_ERROR                   => 'Nameserver {ns}/{address}...',
         QNAME_CASE_INSENSITIVE     => 'Nameserver {ns}/{address} does not preserve original case of queried names.',
         QNAME_CASE_SENSITIVE       => 'Nameserver {ns}/{address} preserves original case of queried names.',
         QUERY_DROPPED              => 'Nameserver {ns}/{address} dropped AAAA query.',
         SAME_SOURCE_IP             => 'All nameservers reply with same IP used to query them.',
+        UNKNOWN_OPTION_CODE        => 'Nameserver {ns}/{address} accepts an unknown EDNS option-code.',
+        UNSUPPORTED_EDNS_VER       => 'Nameserver {ns}/{address} accepts an unsupported EDNS version.',
         UPWARD_REFERRAL            => 'Nameserver {ns}/{address} returns an upward referral.',
         UPWARD_REFERRAL_IRRELEVANT => 'Upward referral tests skipped for root zone.',
+        Z_FLAGS_NOTCLEAR           => 'Nameserver {ns}/{address}...',
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Nameserver.pm]:
```
Nameserver {ns}/{address} has one or more unknown EDNS Z flag bits set.
```

File: lib/Zonemaster/Engine/Test/Nameserver.pm [hunk 12]
```diff
@@ -167,16 +211,22 @@ sub translation {
...
         NO_RECURSOR                => 'Nameserver {ns}/{address} is not a recursor.',
         NO_RESOLUTION              => 'No nameservers succeeded to resolve to an IP address.',
         NO_RESPONSE                => 'No response from {ns}/{address} asking for {dname}.',
         NO_UPWARD_REFERRAL         => 'None of the following nameservers returns an upward referral : {names}.',
+        NS_ERROR                   => 'Nameserver {ns}/{address}...',
         QNAME_CASE_INSENSITIVE     => 'Nameserver {ns}/{address} does not preserve original case of queried names.',
         QNAME_CASE_SENSITIVE       => 'Nameserver {ns}/{address} preserves original case of queried names.',
         QUERY_DROPPED              => 'Nameserver {ns}/{address} dropped AAAA query.',
         SAME_SOURCE_IP             => 'All nameservers reply with same IP used to query them.',
+        UNKNOWN_OPTION_CODE        => 'Nameserver {ns}/{address} accepts an unknown EDNS option-code.',
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Nameserver.pm]: What we test for is the existence of the unknown OPTION-CODE in the response.
```
Nameserver {ns}/{address} responds with an unknown ENDS OPTION-CODE.
```

File: lib/Zonemaster/Engine/Test/Nameserver.pm (hunk 11 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Nameserver.pm]: Fixed

File: lib/Zonemaster/Engine/Test/Nameserver.pm (hunk 12 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Nameserver.pm]: Fixed.

**Comment by sandoche2k** [comment]
@matsduf do you approve?

File: lib/Zonemaster/Engine/Profile.pm [hunk 13]
```diff
@@ -516,6 +519,12 @@ The number of times a query is sent before we give up. Default 2.
 A boolean. If false, UDP queries that get responses with the C<TC>
 flag set will be automatically resent over TCP. Default false.
 
+=head2 resolver.defaults.fallback
+
+A boolean. If false, UDP queries that get responses with the C<TC>
+flag set will be automatically resent over TCP or using EDNS. Default 
+true.
```
[the remark this answers is not in the document]
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Profile.pm]: See ldns_resolver_send_pkt function in resolver.c (https://git.nlnetlabs.nl/ldns/tree/resolver.c?h=release-1.7.0&id=470934b7d8cb16e75de4da9fc4da7ab7ddc58cd1#n1200) for more details

**Pull request merged**